_episodes/l1-04-abstractions.md

```diff
@@ -248,11 +245,11 @@ def compute_result(a, b, experiment) -> np.ndarray:
 It's often a **bad** idea for functions to modify their arguments.
```

Comment:
I think we should moderate a bit this strong statement. Very often, methods are designed to modify the state of the object, which is one of the input arguments - self -, and there might be cases where modifying the input is more convenient. Fortran does that continuously in subroutines. So, yes, it is in general desirable but calling it a "**bad** idea" is not necessary true.

Reply:
This is a bit of a tricky one. I think it is fair to say that this is generally a bad idea in Python. Methods are arguably a different case to standard functions. Outside of methods I'd usually only consider modifying an input if its absolutely required for optimisation reasons. That self is an argument of methods is more of a syntatic quirk of Python.

But, as you point out, in some languages this is how they work by design. Other languages, like Haskell, actively forbid this (thou functions shalt be pure!). So I will update to make the advice more generally applicable.